File: sdk/dart/persona_agent_flutter/lib/src/utils/tool_presentation.dart
import 'dart:convert';

import 'package:flutter/material.dart';

/// Humanized titles + semantic icons + result parsing for the built-in
/// tools every deepagents-based Persona agent ships with (file ops, web/KB
/// search, grep, todos, subagents). Ported from `sdk/ui/src/utils/
/// toolPresentation.ts`, which was itself ported from persona.hasanraiyan.me's
/// own frontend, so a tool call renders with the same per-tool presentation
/// across web, Flutter, and the reference site.

Map<String, dynamic> _asMap(Object? value) => value is Map<String, dynamic> ? value : const {};

String _firstString(Map<String, dynamic> args, List<String> keys) {
  for (final key in keys) {
    final value = args[key];
    if (value is String && value.isNotEmpty) return value;
  }
  return '';
}

bool isWebSearchTool(String name) {
  final n = name.toLowerCase();
  return n == 'search_web' || n.contains('google') || n.startsWith('tavily');
}

bool isKbSearchTool(String name) {
  final n = name.toLowerCase();
  return !isWebSearchTool(name) && (n == 'search_knowledge_base' || (n.startsWith('search_') && n != 'search_web'));
}

bool isKbListSourcesTool(String name) {
  final n = name.toLowerCase();
  return n == 'list_knowledge_base_sources' || n.startsWith('list_sources_');
}

bool isGrepTool(String name) => name.toLowerCase().contains('grep');

bool isReadFileTool(String name) {
  final n = name.toLowerCase();
  return n == 'read_file' || n == 'view_file' || n.contains('read_file') || n.contains('view_file');
}

bool isLsTool(String name) {
  final n = name.toLowerCase();
  return n == 'ls' || n == 'list_dir' || n == 'list_directory' || n.contains('list_dir') || n.contains('list_directory');
}

bool isFileWriteTool(String name) => name.toLowerCase() == 'write_file';

bool isFileEditTool(String name) => name.toLowerCase() == 'edit_file';

bool isSubagentTool(String name) => name.toLowerCase() == 'task';

bool isTodoTool(String name) => name.toLowerCase().contains('todo');

IconData getToolIcon(String toolName) {
  final n = toolName.toLowerCase();
  if (isWebSearchTool(toolName)) return Icons.public;
  if (isKbSearchTool(toolName) || isKbListSourcesTool(toolName)) return Icons.menu_book_rounded;
  if (isGrepTool(toolName)) return Icons.search;
  if (n.contains('todo')) return Icons.checklist_rounded;
  if (isSubagentTool(toolName)) return Icons.smart_toy_outlined;
  if (isReadFileTool(toolName) || isLsTool(toolName) || isFileWriteTool(toolName) || isFileEditTool(toolName) || n.contains('file')) {
    return Icons.description_outlined;
  }
  return Icons.build_outlined;
}

String queryFromArgs(Map<String, dynamic> args) => _firstString(args, ['query', 'q', 'search_query', 'text', 'input']);

String getToolTitle(String toolName, Map<String, dynamic> args, {required bool done}) {
  final query = queryFromArgs(args);

  if (isWebSearchTool(toolName)) {
    if (query.isNotEmpty) return done ? 'Searched the web for "$query"' : 'Searching the web for "$query"';
    return done ? 'Searched the web' : 'Searching the web';
  }

  if (isKbSearchTool(toolName) || isKbListSourcesTool(toolName)) {
    final kbName = _firstString(args, ['knowledgeBaseName']).isEmpty ? 'Knowledge Base' : args['knowledgeBaseName'] as String;
    final isSearch = !isKbListSourcesTool(toolName);
    final kbQuery = _firstString(args, ['query']).isEmpty ? query : args['query'] as String;
    if (isSearch) {
      if (kbQuery.isNotEmpty) {
        return done ? 'Searched knowledge base "$kbName" for "$kbQuery"' : 'Searching knowledge base "$kbName" for "$kbQuery"';
      }
      return done ? 'Searched knowledge base "$kbName"' : 'Searching knowledge base "$kbName"';
    }
    return done ? 'Listed documents in "$kbName"' : 'Listing documents in "$kbName"';
  }

  if (toolName.toLowerCase().contains('todo')) {
    return done ? 'Updated the plan' : 'Updating the plan';
  }

  if (isReadFileTool(toolName)) return done ? 'Read file' : 'Reading file';
  if (isLsTool(toolName)) return done ? 'Listed directory' : 'Listing directory';

  if (isSubagentTool(toolName)) {
    final subagentType = _firstString(args, ['subagent_type']);
    final label = subagentType.isEmpty ? 'subagent' : '$subagentType subagent';
    return done ? 'Ran $label' : 'Running $label';
  }

  if (toolName.toLowerCase().contains('file') || toolName.toLowerCase() == 'glob') {
    return done ? 'Updated files' : 'Working with files';
  }

  return toolName
      .split(RegExp(r'[_\-\s]'))
      .where((w) => w.isNotEmpty)
      .map((w) => w[0].toUpperCase() + w.substring(1))
      .join(' ');
}

String getDomain(String url) {
  try {
    final host = Uri.parse(url).host;
    return host.startsWith('www.') ? host.substring(4) : host;
  } catch (_) {
    return url;
  }
}

class PersonaSearchResult {
  const PersonaSearchResult({this.title, this.url});
  final String? title;
  final String? url;
}

List<PersonaSearchResult> searchResults(Object? result) {
  final decoded = _tryDecode(result);
  final list = decoded is List
      ? decoded
      : (decoded is Map && decoded['results'] is List)
      ? decoded['results'] as List
      : const [];
  return list.whereType<Map>().map((m) {
    final map = m.cast<String, dynamic>();
    return PersonaSearchResult(title: map['title'] as String?, url: map['url'] as String?);
  }).toList();
}

class PersonaLsEntry {
  const PersonaLsEntry({required this.name, required this.isDir});
  final String name;
  final bool isDir;
}

List<PersonaLsEntry> parseLsResults(String? resultText) {
  if (resultText == null || resultText.isEmpty) return const [];
  final decoded = _tryDecode(resultText);
  if (decoded is List) {
    return decoded.map((item) {
      if (item is String) {
        final isDir = item.endsWith('/') || item.contains('(directory)');
        return PersonaLsEntry(name: item.replaceAll('(directory)', '').trim(), isDir: isDir);
      }
      final map = _asMap(item);
      return PersonaLsEntry(
        name: (map['name'] ?? map['path'] ?? '').toString(),
        isDir: map['isDir'] == true || map['is_dir'] == true || map['isDirectory'] == true,
      );
    }).toList();
  }

  return resultText
      .split('\n')
      .map((l) => l.trim())
      .where((l) => l.isNotEmpty)
      .map((line) {
        final isDir = line.endsWith('/') || line.toLowerCase().contains('(directory)') || line.toLowerCase().contains('(dir)');
        final name = line.replaceAll(RegExp(r'\(directory\)', caseSensitive: false), '').replaceAll(RegExp(r'\(dir\)', caseSensitive: false), '').trim();
        return PersonaLsEntry(name: name, isDir: isDir);
      })
      .toList();
}

class PersonaGrepMatch {
  const PersonaGrepMatch({required this.file, required this.line, required this.content});
  final String file;
  final int line;
  final String content;
}

List<PersonaGrepMatch> parseGrepResults(String? resultText) {
  if (resultText == null || resultText.isEmpty) return const [];
  final decoded = _tryDecode(resultText);
  final matches = decoded is List ? decoded : (decoded is Map && decoded['matches'] is List) ? decoded['matches'] as List : null;
  if (matches != null) {
    return matches.whereType<Map>().map((m) {
      final map = m.cast<String, dynamic>();
      final file = (map['Filename'] ?? map['filename'] ?? map['file'] ?? map['path'] ?? '').toString();
      final line = int.tryParse((map['LineNumber'] ?? map['lineNumber'] ?? map['line'] ?? 0).toString()) ?? 0;
      final content = (map['LineContent'] ?? map['lineContent'] ?? map['content'] ?? '').toString();
      return PersonaGrepMatch(file: file, line: line, content: content);
    }).where((m) => m.file.isNotEmpty).toList();
  }

  final out = <PersonaGrepMatch>[];
  var currentFile = '';
  for (final raw in resultText.split('\n')) {
    final line = raw.trim();
    if (line.isEmpty) continue;
    if (line.endsWith(':') && !RegExp(r'^\d+:').hasMatch(line)) {
      currentFile = line.substring(0, line.length - 1).trim();
    } else if (currentFile.isNotEmpty) {
      final match = RegExp(r'^(\d+):(.*)$').firstMatch(line);
      if (match != null) {
        out.add(PersonaGrepMatch(file: currentFile, line: int.parse(match.group(1)!), content: match.group(2)!));
      } else {
        out.add(PersonaGrepMatch(file: currentFile, line: 0, content: line));
      }
    }
  }
  return out;
}

enum DiffRowType { context, add, remove }

class DiffRow {
  const DiffRow(this.type, this.line);
  final DiffRowType type;
  final String line;
}

/// Bounded LCS diff — falls back to naive remove-then-add when either side
/// is too large for O(n*m) DP to stay fast.
List<DiffRow> computeLineDiff(List<String> oldLines, List<String> newLines) {
  final n = oldLines.length;
  final m = newLines.length;
  if (n * m > 250000) {
    return [
      ...oldLines.map((l) => DiffRow(DiffRowType.remove, l)),
      ...newLines.map((l) => DiffRow(DiffRowType.add, l)),
    ];
  }

  final dp = List.generate(n + 1, (_) => List<int>.filled(m + 1, 0));
  for (var i = n - 1; i >= 0; i--) {
    for (var j = m - 1; j >= 0; j--) {
      dp[i][j] = oldLines[i] == newLines[j] ? dp[i + 1][j + 1] + 1 : (dp[i + 1][j] > dp[i][j + 1] ? dp[i + 1][j] : dp[i][j + 1]);
    }
  }

  final rows = <DiffRow>[];
  var i = 0, j = 0;
  while (i < n && j < m) {
    if (oldLines[i] == newLines[j]) {
      rows.add(DiffRow(DiffRowType.context, oldLines[i]));
      i++;
      j++;
    } else if (dp[i + 1][j] >= dp[i][j + 1]) {
      rows.add(DiffRow(DiffRowType.remove, oldLines[i]));
      i++;
    } else {
      rows.add(DiffRow(DiffRowType.add, newLines[j]));
      j++;
    }
  }
  while (i < n) {
    rows.add(DiffRow(DiffRowType.remove, oldLines[i++]));
  }
  while (j < m) {
    rows.add(DiffRow(DiffRowType.add, newLines[j++]));
  }
  return rows;
}

class PersonaDiffStats {
  const PersonaDiffStats({required this.added, required this.removed});
  final int added;
  final int removed;
}

PersonaDiffStats? computeFileDiffStats(String toolName, Map<String, dynamic> args) {
  if (isFileWriteTool(toolName)) {
    final content = args['content'];
    if (content is! String) return null;
    return PersonaDiffStats(added: content.split('\n').length, removed: 0);
  }

  if (isFileEditTool(toolName)) {
    final oldString = args['old_string'];
    final newString = args['new_string'];
    if (oldString is! String && newString is! String) return null;
    final rows = computeLineDiff(
      (oldString as String? ?? '').split('\n'),
      (newString as String? ?? '').split('\n'),
    );
    return PersonaDiffStats(
      added: rows.where((r) => r.type == DiffRowType.add).length,
      removed: rows.where((r) => r.type == DiffRowType.remove).length,
    );
  }

  return null;
}

String getFilePathFromArgs(Map<String, dynamic> args) =>
    _firstString(args, ['file_path', 'filePath', 'path', 'filename', 'fileName', 'targetFile', 'target_file']);

Object? _tryDecode(Object? value) {
  if (value is! String) return value;
  try {
    return jsonDecode(value);
  } catch (_) {
    return null;
  }
}

File: sdk/ui/src/utils/toolPresentation.ts
import type { ComponentType } from 'react';
import { BookText, Bot, FileText, Globe, ListTodo, Search, Wrench } from 'lucide-react';

/**
 * Humanized titles + semantic icons + result parsing for the built-in tools
 * every deepagents-based persona ships with (file ops, web/KB search, grep,
 * todos, subagents). Ported from persona.hasanraiyan.me's own frontend
 * (frontend/src/components/agents/agui/utils.js's toolTitle/subToolIcon and
 * the tool-cards/*.jsx result parsers) so sdk/ui's generic tool trace gets
 * the same per-tool presentation instead of a raw tool name + JSON dump.
 */

function asRecord(value: unknown): Record<string, unknown> {
  return typeof value === 'object' && value !== null ? (value as Record<string, unknown>) : {};
}

function firstString(args: Record<string, unknown>, keys: string[]): string {
  for (const key of keys) {
    const value = args[key];
    if (typeof value === 'string' && value) return value;
  }
  return '';
}

export function isWebSearchTool(name: string): boolean {
  const n = name.toLowerCase();
  return n === 'search_web' || n.includes('google') || n.startsWith('tavily');
}

export function isKbSearchTool(name: string): boolean {
  const n = name.toLowerCase();
  return !isWebSearchTool(name) && (n === 'search_knowledge_base' || (n.startsWith('search_') && n !== 'search_web'));
}

export function isKbListSourcesTool(name: string): boolean {
  const n = name.toLowerCase();
  return n === 'list_knowledge_base_sources' || n.startsWith('list_sources_');
}

export function isGrepTool(name: string): boolean {
  return name.toLowerCase().includes('grep');
}

export function isReadFileTool(name: string): boolean {
  const n = name.toLowerCase();
  return n === 'read_file' || n === 'view_file' || n.includes('read_file') || n.includes('view_file');
}

export function isLsTool(name: string): boolean {
  const n = name.toLowerCase();
  return n === 'ls' || n === 'list_dir' || n === 'list_directory' || n.includes('list_dir') || n.includes('list_directory');
}

export function isFileWriteTool(name: string): boolean {
  return name.toLowerCase() === 'write_file';
}

export function isFileEditTool(name: string): boolean {
  return name.toLowerCase() === 'edit_file';
}

export function isSubagentTool(name: string): boolean {
  return name.toLowerCase() === 'task';
}

export function getToolIcon(toolName: string): ComponentType<{ className?: string }> {
  const n = toolName.toLowerCase();
  if (isWebSearchTool(toolName)) return Globe;
  if (isKbSearchTool(toolName) || isKbListSourcesTool(toolName)) return BookText;
  if (isGrepTool(toolName)) return Search;
  if (n.includes('todo')) return ListTodo;
  if (isSubagentTool(toolName)) return Bot;
  if (isReadFileTool(toolName) || isLsTool(toolName) || isFileWriteTool(toolName) || isFileEditTool(toolName) || n.includes('file')) {
    return FileText;
  }
  return Wrench;
}

export function queryFromArgs(args: unknown): string {
  const record = asRecord(args);
  const value = firstString(record, ['query', 'q', 'search_query', 'text', 'input']);
  return value;
}

export function getToolTitle(toolName: string, args: unknown, status: 'running' | 'completed'): string {
  const done = status === 'completed';
  const query = queryFromArgs(args);
  const record = asRecord(args);

  if (isWebSearchTool(toolName)) {
    if (query) return done ? `Searched the web for "${query}"` : `Searching the web for "${query}"`;
    return done ? 'Searched the web' : 'Searching the web';
  }

  if (isKbSearchTool(toolName) || isKbListSourcesTool(toolName)) {
    const kbName = firstString(record, ['knowledgeBaseName']) || 'Knowledge Base';
    const isSearch = !isKbListSourcesTool(toolName);
    const kbQuery = firstString(record, ['query']) || query;
    if (isSearch) {
      if (kbQuery) return done ? `Searched knowledge base "${kbName}" for "${kbQuery}"` : `Searching knowledge base "${kbName}" for "${kbQuery}"`;
      return done ? `Searched knowledge base "${kbName}"` : `Searching knowledge base "${kbName}"`;
    }
    return done ? `Listed documents in "${kbName}"` : `Listing documents in "${kbName}"`;
  }

  if (toolName.toLowerCase().includes('todo')) {
    return done ? 'Updated the plan' : 'Updating the plan';
  }

  if (isReadFileTool(toolName)) {
    return done ? 'Read file' : 'Reading file';
  }

  if (isLsTool(toolName)) {
    return done ? 'Listed directory' : 'Listing directory';
  }

  if (isSubagentTool(toolName)) {
    const subagentType = firstString(record, ['subagent_type']);
    const label = subagentType ? `${subagentType} subagent` : 'subagent';
    return done ? `Ran ${label}` : `Running ${label}`;
  }

  if (toolName.toLowerCase().includes('file') || toolName.toLowerCase() === 'glob') {
    return done ? 'Updated files' : 'Working with files';
  }

  return toolName
    .split(/[_\-\s]/)
    .filter(Boolean)
    .map((word) => word[0].toUpperCase() + word.slice(1))
    .join(' ');
}

export function getDomain(url: string): string {
  try {
    return new URL(url).hostname.replace(/^www\./, '');
  } catch {
    return url;
  }
}

export interface PersonaSearchResult {
  title?: string;
  url?: string;
}

export function searchResults(result: unknown): PersonaSearchResult[] {
  if (Array.isArray(result)) return result as PersonaSearchResult[];
  const record = asRecord(result);
  if (Array.isArray(record.results)) return record.results as PersonaSearchResult[];
  return [];
}

export interface PersonaLsEntry {
  name: string;
  isDir: boolean;
}

export function parseLsResults(result: unknown): PersonaLsEntry[] {
  if (Array.isArray(result)) {
    return (result as unknown[]).map((item) => {
      if (typeof item === 'string') {
        const isDir = item.endsWith('/') || item.includes('(directory)');
        return { name: item.replace(/\(directory\)/g, '').trim(), isDir };
      }
      const record = asRecord(item);
      return {
        name: String(record.name ?? record.path ?? ''),
        isDir: Boolean(record.isDir ?? record.is_dir ?? record.isDirectory),
      };
    });
  }

  if (typeof result !== 'string' || !result) return [];

  return result
    .split('\n')
    .map((line) => line.trim())
    .filter(Boolean)
    .map((line) => {
      const isDir = line.endsWith('/') || line.toLowerCase().includes('(directory)') || line.toLowerCase().includes('(dir)');
      const name = line.replace(/\(directory\)/gi, '').replace(/\(dir\)/gi, '').trim();
      return { name, isDir };
    });
}

export interface PersonaGrepMatch {
  file: string;
  line: number;
  content: string;
}

export function parseGrepResults(result: unknown): PersonaGrepMatch[] {
  const record = Array.isArray(result) ? { matches: result } : asRecord(result);
  const matches = Array.isArray(record.matches) ? record.matches : Array.isArray(record.results) ? record.results : null;
  if (Array.isArray(matches)) {
    return matches
      .map((m) => {
        const row = asRecord(m);
        return {
          file: String(row.Filename ?? row.filename ?? row.file ?? row.path ?? ''),
          line: Number(row.LineNumber ?? row.lineNumber ?? row.line ?? 0),
          content: String(row.LineContent ?? row.lineContent ?? row.content ?? ''),
        };
      })
      .filter((m) => m.file);
  }

  if (typeof result !== 'string' || !result) return [];

  const out: PersonaGrepMatch[] = [];
  const lines = result.split('\n');
  let currentFile = '';
  for (const raw of lines) {
    const line = raw.trim();
    if (!line) continue;
    if (line.endsWith(':') && !/^\d+:/.test(line)) {
      currentFile = line.slice(0, -1).trim();
    } else if (currentFile) {
      const match = line.match(/^(\d+):(.*)$/);
      if (match) {
        out.push({ file: currentFile, line: parseInt(match[1], 10), content: match[2] });
      } else {
        out.push({ file: currentFile, line: 0, content: line });
      }
    }
  }

  if (out.length === 0) {
    return lines
      .map((l) => l.trim())
      .filter((l) => l && (l.startsWith('/') || l.includes('.') || l.includes('\\')))
      .map((l) => ({ file: l, line: 0, content: '' }));
  }

  return out;
}

export type DiffRow = { type: 'context' | 'add' | 'remove'; line: string };

// Bounded LCS diff: falls back to naive remove-then-add when either side is
// too large for O(n*m) DP to stay fast (matches the reference's 250k cap).
export function computeLineDiff(oldLines: string[], newLines: string[]): DiffRow[] {
  const n = oldLines.length;
  const m = newLines.length;
  if (n * m > 250000) {
    return [
      ...oldLines.map((line) => ({ type: 'remove' as const, line })),
      ...newLines.map((line) => ({ type: 'add' as const, line })),
    ];
  }

  const dp: number[][] = Array.from({ length: n + 1 }, () => new Array(m + 1).fill(0));
  for (let i = n - 1; i >= 0; i--) {
    for (let j = m - 1; j >= 0; j--) {
      dp[i][j] = oldLines[i] === newLines[j] ? dp[i + 1][j + 1] + 1 : Math.max(dp[i + 1][j], dp[i][j + 1]);
    }
  }

  const rows: DiffRow[] = [];
  let i = 0;
  let j = 0;
  while (i < n && j < m) {
    if (oldLines[i] === newLines[j]) {
      rows.push({ type: 'context', line: oldLines[i] });
      i++;
      j++;
    } else if (dp[i + 1][j] >= dp[i][j + 1]) {
      rows.push({ type: 'remove', line: oldLines[i] });
      i++;
    } else {
      rows.push({ type: 'add', line: newLines[j] });
      j++;
    }
  }
  while (i < n) rows.push({ type: 'remove', line: oldLines[i++] });
  while (j < m) rows.push({ type: 'add', line: newLines[j++] });
  return rows;
}

export interface PersonaDiffStats {
  added: number;
  removed: number;
}

// Line counts for the row's "+N -N" diffstat badge. Returns null when the
// args aren't parseable (yet) — the caller falls back to the generic panel.
export function computeFileDiffStats(toolName: string, args: unknown): PersonaDiffStats | null {
  const record = asRecord(args);

  if (isFileWriteTool(toolName)) {
    if (typeof record.content !== 'string') return null;
    return { added: record.content.split('\n').length, removed: 0 };
  }

  if (isFileEditTool(toolName)) {
    if (typeof record.old_string !== 'string' && typeof record.new_string !== 'string') return null;
    const rows = computeLineDiff(
      String(record.old_string ?? '').split('\n'),
      String(record.new_string ?? '').split('\n')
    );
    return {
      added: rows.filter((r) => r.type === 'add').length,
      removed: rows.filter((r) => r.type === 'remove').length,
    };
  }

  return null;
}

export function getFilePathFromArgs(args: unknown): string {
  const record = asRecord(args);
  return firstString(record, ['file_path', 'filePath', 'path', 'filename', 'fileName', 'targetFile', 'target_file']);
}
